plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php

```diff
@@ -137,6 +176,18 @@ public function create( &$product ) {
 
 		if ( $id && ! is_wp_error( $id ) ) {
 			$product->set_id( $id );
+			$sku = $product->get_sku();
+
+			/**
+			 * If SKU is already under processing aka Duplicate SKU
+			 * because of concurrent requests, then we should not proceed
+			 * Delete the product and throw an exception.
+			 */
+			if ( ! empty( $sku ) && ! $this->obtain_lock_on_sku_for_concurrent_requests( $product ) ) {
+				$product->delete( true );
+				// translators: 1: SKU, 2: Product Id.
+				throw new Exception( esc_html( sprintf( __( 'The SKU (%1$s) you are trying to insert with Product Id (%2$s) is already under processing', 'woocommerce' ), $sku, $id ) ) );
```

Comment:
Sorry I didn't think of this earlier, but it occurs to me that throwing an exception in this existing public method might be breaking backcompat, because other codebases that use the method wouldn't be wrapping it in a try/catch (thus risking an "uncaught exception" fatal error). In fact, there is at least one other place in this codebase besides the REST CRUD controller that is using it that this PR doesn't update with a try/catch ([example](https://github.com/woocommerce/woocommerce/blob/f56d3e3ee04a47f750e71bc8c09c4152c293052d/plugins/woocommerce/includes/abstracts/abstract-wc-product.php#L1435-L1435)).

I don't have a better idea off the top of my head for passing the result of the SKU lock check on to the caller, but I do think it's worth considering the risks here. Wdyt?

Reply:
Being discussed here: p1717092314975919-slack-C0E1AV8T0

Reply:
Resolved as part of https://github.com/woocommerce/woocommerce/pull/47476/commits/5b596d822472d79cecfb1449be3497b40fd0a2af